lnwallet/btcwallet/btcwallet.go

```diff
@@ -429,47 +429,29 @@ func (b *BtcWallet) ListUnspentWitness(minConfs, maxConfs int32) (
 // network (either in the mempool or chain) no error will be returned.
 func (b *BtcWallet) PublishTransaction(tx *wire.MsgTx) error {
 	if err := b.wallet.PublishTransaction(tx); err != nil {
-		switch b.chain.(type) {
-		case *chain.RPCClient:
-			if strings.Contains(err.Error(), "already spent") {
-				// Output was already spent.
-				return lnwallet.ErrDoubleSpend
-			}
-			if strings.Contains(err.Error(), "already been spent") {
-				// Output was already spent.
-				return lnwallet.ErrDoubleSpend
-			}
-			if strings.Contains(err.Error(), "orphan transaction") {
-				// Transaction is spending either output that
-				// is missing or already spent.
-				return lnwallet.ErrDoubleSpend
-			}
 
-		case *chain.BitcoindClient:
-			if strings.Contains(err.Error(), "txn-mempool-conflict") {
-				// Output was spent by other transaction
-				// already in the mempool.
-				return lnwallet.ErrDoubleSpend
-			}
-			if strings.Contains(err.Error(), "insufficient fee") {
-				// RBF enabled transaction did not have enough fee.
-				return lnwallet.ErrDoubleSpend
-			}
-			if strings.Contains(err.Error(), "Missing inputs") {
-				// Transaction is spending either output that
-				// is missing or already spent.
-				return lnwallet.ErrDoubleSpend
-			}
+		// If we failed to publish the transaction, check whether we
+		// got an error of known type.
+		var (
+			doubleSpendErr *base.ErrDoubleSpend
+			replacementErr *base.ErrReplacement
+		)
 
-		case *chain.NeutrinoClient:
-			if strings.Contains(err.Error(), "already spent") {
-				// Output was already spent.
-				return lnwallet.ErrDoubleSpend
-			}
+		switch {
+		// If the wallet reports a double spend, convert it to our
+		// internal ErrDoubleSpend and return.
+		case errors.As(err, &doubleSpendErr):
```

Comment:
I think we can just get by with using `Is` here since we're not actually using any of the fields in the populated error (the double (?) pointer receiver). However, we may need to do a type assertion if we want to allow things to be built with go 1.12. Alternatively, I _think_ we can still use the `xerrors` package.

Reply:
Very nice clean up here though, something we've been wanting to do for some time now.

Reply:
Changed to regular type switch, to support go 1.12: https://github.com/lightningnetwork/lnd/pull/3510/commits/f1b5809b0a589ee9888fb194603275bafe50233e